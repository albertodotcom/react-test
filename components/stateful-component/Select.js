import React from 'react';
import PropTypes from 'prop-types';
import { find, compose, map, prop, join, reduce } from 'ramda';

const getSelectedTextByValues = (options, selected) => compose(
  join(', '),
  map(prop('text')),
  reduce((acc, id) => {
    const option = find(o => id === o.id, options);
    return option ? [...acc, option] : acc;
  }, []),
)(selected);

class Dropdown extends React.Component {
  constructor(props) {
    super(props);

    this.state = {
      selected: props.selected,
    };

    this.onChange = this.onChange.bind(this);
    this.onReset = this.onReset.bind(this);
    this.onSubmit = this.onSubmit.bind(this);
  }

  onChange(e) {
    this.setState({
      selected: [...this.state.selected, e.target.value],
    });
  }

  onReset() {
    this.setState({
      selected: [],
    });
  }

  onSubmit() {
    this.props.onSubmit(this.state.selected);
  }

  render() {
    const selectedOptions = getSelectedTextByValues(this.props.options, this.state.selected);
    return (
      <form>
        <div className="select-options t-selected-options">{selectedOptions}</div>
        <select className="select t-select" onChange={this.onChange}>
          {this.props.options.map(o =>
            <option className="t-option" key={o.id} value={o.id}>{o.text}</option>,
          )}
        </select>
        <div>
          <button type="reset" className="t-reset" onClick={this.onReset}>Clear</button>
          <button type="submit" className="t-submit" onClick={this.onSubmit}>Submit</button>
        </div>
      </form>
    );
  }
}

Dropdown.propTypes = {
  options: PropTypes.arrayOf(PropTypes.shape({
    id: PropTypes.string.isRequired,
    text: PropTypes.string.isRequired,
  })).isRequired,
  selected: PropTypes.arrayOf(PropTypes.string),
  onSubmit: PropTypes.func.isRequired,
};

Dropdown.defaultProps = {
  selected: [],
};

export default Dropdown;
