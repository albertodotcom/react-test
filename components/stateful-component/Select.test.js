/* eslint-env jest */
import React from 'react';
import { shallow } from 'enzyme';

import Select from './Select';

describe('Select', () => {
  const defaultProps = {
    options: [
      { id: '8', text: 'Conte' },
      { id: '10', text: 'Del Piero' },
      { id: '1', text: 'Buffon' },
    ],
    onSubmit: () => { },
  };

  describe('element testing', () => {
    it('renders the passed in options', () => {
      const options = [
        { id: '10', text: 'Del Piero' },
        { id: '21', text: 'Dybala' },
      ];

      const wrapper = shallow(
        <Select
          {...defaultProps}
          options={options}
        />);

      const optionTags = wrapper.find('.t-option');
      expect(optionTags.length).toBe(2);
      expect(optionTags.at(0).text()).toBe(options[0].text);
      expect(optionTags.at(1).text()).toBe(options[1].text);
      expect(wrapper.find('.t-selected-options').text()).toBe('');
    });

    it('renders the selected option', () => {
      const wrapper = shallow(
        <Select
          {...defaultProps}
          selected={['10']}
        />);

      expect(wrapper.find('.t-selected-options').text()).toBe('Del Piero');

      wrapper.find('.t-select').simulate('change', { target: { value: '1' } });
      expect(wrapper.find('.t-selected-options').text()).toBe('Del Piero, Buffon');
    });

    describe('DONTS', () => {
      it('displays the selected items by changing state', () => {
        const wrapper = shallow(
          <Select
            {...defaultProps}
            selected={['10']}
          />);

        expect(wrapper.find('.t-selected-options').text()).toBe('Del Piero');

        wrapper.setState({ selected: ['1', '10'] });
        expect(wrapper.find('.t-selected-options').text()).toBe('Buffon, Del Piero');
      });
    });
  });

  describe('snapshot testing', () => {
    it('user can select options', () => {
      const onSubmit = jest.fn();
      const options = [
        { id: '8', text: 'Conte' },
        { id: '10', text: 'Del Piero' },
        { id: '1', text: 'Buffon' },
        { id: '21', text: 'Pirlo' },
      ];

      const wrapper = shallow(
        <Select
          {...defaultProps}
          options={options}
          selected={['1', '21']}
          onSubmit={onSubmit}
        />);

      expect(wrapper).toMatchSnapshot();

      wrapper.find('.t-select').simulate('change', { target: { value: '10' } });
      expect(wrapper).toMatchSnapshot();

      wrapper.find('.t-submit').simulate('click');

      expect(onSubmit.mock.calls.length).toBe(1);
      expect(onSubmit).toBeCalledWith(['1', '21', '10']);

      wrapper.find('.t-reset').simulate('click');
      expect(wrapper).toMatchSnapshot();
    });
  });


  describe('interactions', () => {
    it('clears the selection when clicking on reset', () => {
      const wrapper = shallow(
        <Select
          {...defaultProps}
          selected={['10']}
        />);

      wrapper.find('.t-reset').simulate('click');

      expect(wrapper.find('.t-selected-options').text()).toBe('');
    });

    it('returns the new selection on submit', () => {
      const onSubmit = jest.fn();
      const wrapper = shallow(
        <Select
          {...defaultProps}
          selected={['10']}
          onSubmit={onSubmit}
        />);

      wrapper.find('.t-submit').simulate('click');

      expect(onSubmit.mock.calls.length).toBe(1);
      expect(onSubmit).toBeCalledWith(['10']);
    });
  });
});
