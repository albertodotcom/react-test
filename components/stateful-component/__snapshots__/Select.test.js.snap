// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Select snapshot testing selects options 1`] = `
<form>
  <div
    className="select-options t-selected-options"
  >
    Buffon, Pirlo
  </div>
  <select
    className="select t-select"
    onChange={[Function]}
  >
    <option
      className="t-option"
      value="8"
    >
      Conte
    </option>
    <option
      className="t-option"
      value="10"
    >
      Del Piero
    </option>
    <option
      className="t-option"
      value="1"
    >
      Buffon
    </option>
    <option
      className="t-option"
      value="21"
    >
      Pirlo
    </option>
  </select>
  <div>
    <button
      className="t-reset"
      onClick={[Function]}
      type="reset"
    >
      Clear
    </button>
    <button
      className="t-submit"
      onClick={[Function]}
      type="submit"
    >
      Submit
    </button>
  </div>
</form>
`;

exports[`Select snapshot testing selects options 2`] = `
<form>
  <div
    className="select-options t-selected-options"
  >
    Buffon, Pirlo, Del Piero
  </div>
  <select
    className="select t-select"
    onChange={[Function]}
  >
    <option
      className="t-option"
      value="8"
    >
      Conte
    </option>
    <option
      className="t-option"
      value="10"
    >
      Del Piero
    </option>
    <option
      className="t-option"
      value="1"
    >
      Buffon
    </option>
    <option
      className="t-option"
      value="21"
    >
      Pirlo
    </option>
  </select>
  <div>
    <button
      className="t-reset"
      onClick={[Function]}
      type="reset"
    >
      Clear
    </button>
    <button
      className="t-submit"
      onClick={[Function]}
      type="submit"
    >
      Submit
    </button>
  </div>
</form>
`;

exports[`Select snapshot testing selects options 3`] = `
<form>
  <div
    className="select-options t-selected-options"
  />
  <select
    className="select t-select"
    onChange={[Function]}
  >
    <option
      className="t-option"
      value="8"
    >
      Conte
    </option>
    <option
      className="t-option"
      value="10"
    >
      Del Piero
    </option>
    <option
      className="t-option"
      value="1"
    >
      Buffon
    </option>
    <option
      className="t-option"
      value="21"
    >
      Pirlo
    </option>
  </select>
  <div>
    <button
      className="t-reset"
      onClick={[Function]}
      type="reset"
    >
      Clear
    </button>
    <button
      className="t-submit"
      onClick={[Function]}
      type="submit"
    >
      Submit
    </button>
  </div>
</form>
`;
